src/test/java/com/google/devtools/build/lib/blackbox/tests/workspace/PatchApiBlackBoxTest.java

```diff
@@ -0,0 +1,253 @@
+        "load(\":patched_repo.bzl\", \"patched_repo\")",
+        "",
+        "patched_repo(",
+        "    name = \"test\",",
+        "    files = {\"foo.sh\" : \"//:foo.sh\"},",
+        "    patches = [\"//:remove-dragons.patch\"],",
+        String.format("    patch_args = [%s],", patchArgsStr.toString()),
+        (patchTool == null ? "" : String.format("    patch_tool = \"%s\",", patchTool)),
+        "    patch_cmds = [",
+        "      \"find . -name '*.sh' -exec sed -i.orig '1s|#!/usr/bin/env sh$|/bin/sh$|' {} +\",",
+        "      \"chmod u+x ./foo.sh\",",
+        "    ],",
+        (hasPatchCmdsWin ? "    patch_cmds_win = [\"(Get-Content -path foo.sh) -replace '/usr/bin/env sh','/bin/sh' | Set-Content -Path foo.sh\"]," : ""),
+        "    build_file_content =",
+        "    \"\"\"",
+        "filegroup(",
+        "    name = \"foo\",",
+        "    srcs = [\"foo.sh\"],",
+        ")",
+        "    \"\"\"",
+        ")");
+    context().write("BUILD");
+    context().write("foo.sh",
+        "#!/usr/bin/env sh",
+        "",
+        "echo Here be dragons...",
+        "");
+    context().write("remove-dragons.patch",
+        "#!/usr/bin/env sh",
+        "diff --git a/foo.sh b/foo.sh",
+        "index 1f4c41e..9d548ff 100644",
+        "--- a/foo.sh",
+        "+++ b/foo.sh",
+        "@@ -1,3 +1,3 @@",
+        " #!/usr/bin/env sh",
+        "",
+        "-echo Here be dragons...",
+        "+echo New version of foo.sh, no more dangerous animals...",
+        "");
+  }
+
+  @Test
+  public void testPatchApiUsingNativePatch() throws Exception {
+    setUpPatchTestRepo(null, null, true);
+    BuilderRunner bazel = WorkspaceTestUtils.bazel(context()).withFlags("--incompatible_use_native_patch");
+    bazel.build("@test//:foo");
+    assertFooIsPatched(bazel);
+  }
+
+  @Test
+  public void testPatchApiUsingNativePatchFailed() throws Exception {
+    // Using -p2 should cause an error
+    setUpPatchTestRepo(ImmutableList.of("-p2"), null, true);
+    BuilderRunner bazel = WorkspaceTestUtils.bazel(context()).withFlags("--incompatible_use_native_patch").shouldFail();
+    ProcessResult result = bazel.build("@test//:foo");
+    assertThat(result.errString()).contains("Cannot determine file name with strip = 2 at line 4:\n--- a/foo.sh");
+  }
+
+  @Test
+  public void testPatchApiUsingPatchTool() throws Exception {
+    setUpPatchTestRepo(null, null, true);
+    BuilderRunner bazel = WorkspaceTestUtils.bazel(context()).withFlags("--noincompatible_use_native_patch");
+    if (isWindows()) {
+      // On Windows, we expect no patch tool in PATH after removing MSYS paths from PATH env var.
+      bazel.shouldFail();
+    }
+    ProcessResult result = bazel.build("@test//:foo");
+    if (isWindows()) {
+      assertThat(result.errString()).contains("CreateProcessW(\"patch\" -p1 -i");
+      assertThat(result.errString()).contains("The system cannot find the file specified.");
+    } else {
+      assertFooIsPatched(bazel);
```

Comment:
So, the same question again:)

If we removed "patch" from the default value of the patch_tool attribute, what tool is applied here?
If we still use hardcoded "patch" command, does it make sense to remove the default attribute value to indicate "no operating system tool should be used"? 
I find it very confusing way of communicating what kind of patch tool should be applied.
Is it better to have an explicit attribute, like "use_patch_tool" with default False, and keep patch_tool's default='patch'?

Reply:
I think it's a good idea to adding an extra attribute. Thanks!

Reply:
Hmm, a second thought. This will be an incompatible change, because users now using `patch_tool` will have to also add `use_patch_tool` attribute. This is why one of the test is broken:
https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/3538#6f7a6fdf-eba4-41dc-acbe-4872058ddb9d

Reply:
What about use_patch_tool = True as default for before flag flipping?

Reply:
As we discussed offline, set default value of `use_patch_tool` to `True` will prevent users to migrate with the incompatible flag. We have to stick with the previous way to not break users. I added more comments to document the behavior of `patch_args` and `patch_tool`.